docs/userguide/src/portingguide/code/dummyvm/src/lib.rs

```diff
@@ -0,0 +1,51 @@
+extern crate libc;
+extern crate mmtk;
+#[macro_use]
+extern crate lazy_static;
+
+use mmtk::vm::VMBinding;
+use mmtk::MMTKBuilder;
+use mmtk::MMTK;
+
+pub mod active_plan;
+pub mod api;
+pub mod collection;
+pub mod object_model;
+pub mod reference_glue;
+pub mod scanning;
+
+mod slots;
+
+#[derive(Default)]
+pub struct DummyVM;
+
+// Documentation: https://docs.mmtk.io/api/mmtk/vm/trait.VMBinding.html
+impl VMBinding for DummyVM {
+    type VMObjectModel = object_model::VMObjectModel;
+    type VMScanning = scanning::VMScanning;
+    type VMCollection = collection::VMCollection;
+    type VMActivePlan = active_plan::VMActivePlan;
+    type VMReferenceGlue = reference_glue::VMReferenceGlue;
+    type VMSlot = slots::DummyVMSlot;
+    type VMMemorySlice = slots::DummyVMMemorySlice;
+
+    /// Allowed maximum alignment in bytes.
+    const MAX_ALIGNMENT: usize = 1 << 6;
+}
+
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Mutex;
+
+/// This is used to ensure we initialize MMTk at a specified timing.
+pub static MMTK_INITIALIZED: AtomicBool = AtomicBool::new(false);
+
+lazy_static! {
+    pub static ref BUILDER: Mutex<MMTKBuilder> = Mutex::new(MMTKBuilder::new());
+    pub static ref SINGLETON: MMTK<DummyVM> = {
+        let builder = BUILDER.lock().unwrap();
+        debug_assert!(!MMTK_INITIALIZED.load(Ordering::SeqCst));
+        let ret = mmtk::memory_manager::mmtk_init(&builder);
+        MMTK_INITIALIZED.store(true, std::sync::atomic::Ordering::Relaxed);
+        *ret
+    };
+}
```

Comment:
This may not be the best way to initialize a global `MMTK` singleton.

1.  The `MMTKBuilder` only needs to exist during creating the `MMTK` instance.  It shouldn't be static, but it has to be due to the way `SINGLETON` is initialized.
2.  The `MMTK_INITIALIZED` variable plus the `SINGLETON` lazy static variable seem to reinvent the [`OnceLock`](https://doc.rust-lang.org/std/sync/struct.OnceLock.html) type in the standard library (introduced in 1.70.0).

We used to do that because there were no better alternatives when we were using 1.61 or 1.66.  But I don't want to recommend this way of initialization to the readers as it is not very elegant, although our OpenJDK binding has been doing it this way without problem.

If we are not sure if other ways of initialization will work, I am OK with merging the current code in this PR.  We can start experimenting in the OpenJDK binding, and change the dummy VM in a subsequent PR.  That may take some time and some performance evaluation.

The Ruby binding currently uses `once_cell::sync::OnceCell<RubyBinding>` (and should probably switch to the stdlib, too), and the `RubyBinding` struct has a field `mmtk: &'static MMTK<Ruby>`.  It has been working well so far.

Reply:
> If we are not sure if other ways of initialization will work, I am OK with merging the current code in this PR.

We can introduce a test in this crate to make sure that the initialisation works fine. 

> We can start experimenting in the OpenJDK binding, and change the dummy VM in a subsequent PR.

We don't have to make OpenJDk and other bindings to use the same initialization approach we will use in the dummyvm.

> That may take some time and some performance evaluation.

As long as we don't introduce overhead when we access `SINGLETON`, it should be fine. Initialization just happens once.

Reply:
I am happy to make changes in this PR about the initialisation. If I understand you right, you suggested not to have the `MMTKBuilder` as a static, and try use `OnceLock`/`OnceCell` from the stdlib for `MMTK`.

The reason why `MMTKBuilder` is static is that we need to set options to `MMTKBuilder` from the VM when the VM parses its command line args. How do you deal with this?

Reply:
> The reason why `MMTKBuilder` is static is that we need to set options to `MMTKBuilder` from the VM when the VM parses its command line args. How do you deal with this?

The Ruby binding parses all command line args and store the results in variables in the C program.  Then it creates an `MMTKBuilder` instance, set the options, and instantiate `MMTK`.  See [these lines](https://github.com/mmtk/ruby/blob/310837dbbf6ca8abe662782bdef429369b43bb9f/mmtk_support.c#L283-L309).  The Rust part of the binding exposes several methods for setting options in the `MMTKBuilder` and C code calls them.  The `mmtk_init_binding` function gives the ownership of the builder back to the Rust code, and it instantiates `RubyBinding` and `MMTK`.

Alternatively, the VM binding creates an `MMTKBuilder` before parsing command line args.  During parsing, it sets options in the `MMTKBuilder`.  After parsing, it instantiates `MMTK`.

Reply:
I made some changes. I expose a function to create an MMTk builder, process and mmtk_init require a pointer to the MMTk builder. I expose `InitializeOnce` from mmtk-core, and use it for singleton. I add a test to make sure that the init process works fine.

Can you review this PR again?

Reply:
I made some changes upon that.  See https://github.com/mmtk/mmtk-core/commit/0436112477ac5281c974d173deb1656b18161cbd

I reordered the list of functions and placed `mmtk_create_builder` first, and then option-setting functions, and then `mmtk_init`.  That's probably the order the VM binding will call.

I replaced `InitializeOnce` with `OnceLock` for the `SINGLETON`.  `InitializeOnce` lacks a method that sets the value directly (`OnceLock::set`), but it's probably not worth adding it because `InitializeOnce` is intended for internal use.

I also created a wrapper function `fn mmtk() -> &'static MMTK<DummyVM>` because `OnceLock` does not implement `Deref`.  It calls `OnceLock::get()` which involves a check.  I think it's safer to recommend safe functions instead of faster unsafe ones.  If the binding developers find that it's a performance bottleneck, they can switch to the `once_cell` crate and use the unsafe method `once_cell::sync::OnceCell::get_unchecked` instead.

I changed the `extern "C"` wrapper functions for setting options a little bit.  I feel the name `mmtk_process` is not very straightforward (it's inherited from `OptionSet.process` from JikesRVM).  I categorized the method of setting options into "setting by value" and "setting by string", and each can be either done directly using Rust methods or done with `extern "C"` wrapper functions.  I added a test case which uses `extern "C"` wrappers for them.

Reply:
Do we know if `OnceLock` would be no worse than lazy_static? We should get performance right here for the developers.

The reason that I used `InitializeOnce` is that we maintain `InitializeOnce`, and we know  it is zero-overhead abstraction. Getting this wrong could potentially make everything slow.

Reply:
The source code of `OnceLock` shows that it is implemented just like `InitializeOnce`. (See https://doc.rust-lang.org/src/std/sync/once_lock.rs.html#77-99)

The author of the `once_cell` crate recommended against using `lazy_static!`.  (See https://docs.rs/once_cell/latest/once_cell/#faq)

We can use `cargo expand` to see the macro expansion of `lazy_static!`.  The following snippet is from `cargo expand` in `mmtk-openjdk`.

```rust
#[allow(missing_copy_implementations)]
#[allow(non_camel_case_types)]
#[allow(dead_code)]
pub struct SINGLETON_COMPRESSED {
    __private_field: (),
}
#[doc(hidden)]
pub static SINGLETON_COMPRESSED: SINGLETON_COMPRESSED = SINGLETON_COMPRESSED {
    __private_field: (),
};
impl ::lazy_static::__Deref for SINGLETON_COMPRESSED {
    type Target = MMTK<OpenJDK<true>>;
    fn deref(&self) -> &MMTK<OpenJDK<true>> {
        #[inline(always)]
        fn __static_ref_initialize() -> MMTK<OpenJDK<true>> {
            {
                if !use_compressed_oops() {
                    ::core::panicking::panic("assertion failed: use_compressed_oops()")
                }
                let mut builder = BUILDER.lock().unwrap();
                if !!MMTK_INITIALIZED.load(Ordering::Relaxed) {
                    ::core::panicking::panic(
                        "assertion failed: !MMTK_INITIALIZED.load(Ordering::Relaxed)",
                    )
                }
                set_compressed_pointer_vm_layout(&mut builder);
                let ret = mmtk::memory_manager::mmtk_init(&builder);
                MMTK_INITIALIZED.store(true, std::sync::atomic::Ordering::SeqCst);
                slots::initialize_compressed_oops_base_and_shift();
                *ret
            }
        }
        #[inline(always)]
        fn __stability() -> &'static MMTK<OpenJDK<true>> {
            static LAZY: ::lazy_static::lazy::Lazy<MMTK<OpenJDK<true>>> = ::lazy_static::lazy::Lazy::INIT;
            LAZY.get(__static_ref_initialize)
        }
        __stability()
    }
}
impl ::lazy_static::LazyStatic for SINGLETON_COMPRESSED {
    fn initialize(lazy: &Self) {
        let _ = &**lazy;
    }
}
```

A dereference `&SINGLETON` is expanded to `__stability()` which is `LAZY.get(__static_ref_initialize)` which is defined as:

```rust
pub struct Lazy<T: Sync>(Cell<MaybeUninit<T>>, Once);

impl<T: Sync> Lazy<T> {
    #[allow(deprecated)]
    pub const INIT: Self = Lazy(Cell::new(MaybeUninit::uninit()), ONCE_INIT);

    #[inline(always)]
    pub fn get<F>(&'static self, f: F) -> &T
    where
        F: FnOnce() -> T,
    {
        self.1.call_once(|| {
            self.0.set(MaybeUninit::new(f()));
        });

        // `self.0` is guaranteed to be initialized by this point
        // The `Once` will catch and propagate panics
        unsafe { &*(*self.0.as_ptr()).as_ptr() }
    }
}
```

This is very similar to `OnceLock::get_or_init` and our `InitializeOnce::initialize_once`, except it is using `Cell<MaybeUninit<T>>` instead of `UnsafeCell<MaybeUninit<T>>`.

In theory,

-   `InitializeOnce::get_ref` is the fastest because it involves no checking.  It is as fast as `once_cell::sync::OnceCell::get_unchecked`.
-   `std::sync::OnceLock::get` is as fast as `once_cell::sync::OnceCell::get`, but slower than the above.  They involve a check before getting the reference, but will not attempt to initialize it.
-   Dereferencing a `lazy_static!` should be as fast as `std::sync::OnceLock::get_or_init` and `once_cell:sync::OnceCell::get_or_init`.  They all check for completion before doing initialization.  But in order to call this function, the call site must involve the initializer function and a call to the underlying `once.call_once`, increasing the code complexity, which may affect the code cache friendliness of inlined code.

Reply:
I quickly benchmarked the 4 options with criterion. The time showed `[lower_bound estimate upper_bound]`.
```console
once/once_lock/0        time:   [276.70 ps 277.88 ps 279.29 ps]
once/crate_once_cell/0  time:   [278.82 ps 282.20 ps 285.91 ps]
once/initialize_once/0  time:   [274.92 ps 275.69 ps 276.58 ps]
once/lazy_static/0      time:   [552.82 ps 555.12 ps 557.96 ps]
```

It is clear that we should stop using `lazy_static`. Other options are all good.

Reply:
> I quickly benchmarked the 4 options with criterion.

Which methods did you call, `get` or `get_unchecked`?  Is there any difference between them?

Reply:
I was using `get`.